src/data/docs/abilities.md

```diff
@@ -0,0 +1,578 @@
+
+This is all you need to know to work with existing abilities. See [the typechecking rule for abilities](/docs/language-reference#ability-typechecking) for more detail on the typechecking of abilities.
+
+<a id="creating"></a>
+
+## Creating and handling abilities
+
+Abilities aren't just for working with `IO`. You can create your own abilities! An _ability_ in Unison is introduced by a type declaration which has some set of operations.
+
+### An example ability: `Stream`
+
+A simple example ability is something like `Stream`, which can be used to produce streams of output:
+
+```unison
+ability Stream e where
+  emit : e ->{Stream e} ()
+  -- equivalently
+  -- emit : e -> ()
+
+Stream.range : Nat ->{} Nat ->{Stream Nat} ()
+Stream.range n m =
+  if n >= m then ()
+  else
+    emit n
+    Stream.range (n + 1) m
+```
+
+```ucm
+
+  I found and typechecked these definitions in scratch.u. If you
+  do an `add` or `update`, here's how your codebase would
+  change:
+  
+    ⍟ These new definitions are ok to `add`:
+    
+      ability Stream e
+      Stream.range : Nat -> Nat ->{Stream Nat} ()
+   
+  Now evaluating any watch expressions (lines starting with
+  `>`)... Ctrl+C cancels.
+
+```
+This declaration introduces the function `Stream.emit ` which has exactly the type provided in the ability declaration: `e ->{Stream e} ()` so calling `emit` requires the `Stream e` ability. We call `emit` one of the _operations_ of the ability. In general, an ability declaration can have any number of operations in it.
+
+The `Stream.range` example shows how we can use `emit` to produce streams. (Exercise: try changing `range` to be pure and you'll get an ability check failure.) We can think of the calls to `emit` as suspending the computation and requesting that the `emit` operation be handled by an external piece of code that will interpret the operation and (if it wants) resume the computation.
+
+The operations of an ability are all abstract: the ability declaration just states the signatures of the operations but doesn't say how they are implemented or what they mean. We give meaning to the operations of an ability by interpreting them with a _handler_. Let's now look at an example of a handler, which interprets `Stream` computations by collecting all the emited values into a list. There's a bunch of new things here, which will all be explained!
+
+```unison
+Stream.toList : '{Stream a} () -> [a]
+Stream.toList stream =
+  handle !stream with
+    h : [a] -> Request {Stream a} () -> [a]
+    h acc req = case req of
+      {Stream.emit e -> resume} ->
+        handle resume () with h (snoc acc e)
+      {u} -> acc
+    h []
+
+> Stream.toList '(Stream.range 0 10)
+```
```

Comment:
This was a good suggestion - I played around with this some and couldn't find a way to make it work. I found it was somewhat awkward to try to introduce `handle` blocks separate from introducing a full handler, like it was more confusing and seemed better to hit the reader with a short, complete example and a good signpost that it would be explained in detail.

My original thought was something like `handle 1 + abort with abortHandler defaultVal`. You can sort of intuit what's going on from a few watch expressions, however it also raises a bunch of questions (what is `abortHandler`?  what's its type? why does it have that type? what's `abort`?) 

Aside: I don't really like `abort` as a first example ability, because of its use of an existential return type that prevents the continuation from being resumed.